### PR TITLE
Feature/encore added functionality

### PR DESCRIPTION
**Pull request by bishopbm1**
Added action to query for agent information from orion. Added the needed agent settings into the node definitions. Added get agent id and delete agent functions that use the new agent actions. Also added the ability to delete nodes. Updated mode unmanage action to accept begin and end dates so that unmanage times and be set for future times and can be longer than max set time from the config. Added new tests for all new functions and actions.

### CODE REVIEW

**Comment by LindsayHill** [comment]
Thanks for this @bishopbm1 - much appreciated. Would you mind bumping the version in `pack.yaml`, and adding an entry to CHANGES.md?

@jjm would you mind casting your eye over this since you wrote much of the original pack?

**Comment by jjm** [comment]
@bishopbm1 Thank you for the PR, I'll get this reviewed today.

File: actions/agent_delete.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+entry_point: 'agent_delete.py'
+name: "agent_delete"
+pack: "orion"
+runner_type: "python-script"
+
+parameters:
+  node:
+    type: "string"
+    description: The node to delete in Orion.
+    required: true
```
Comment (jjm) [inline actions/agent_delete.yaml]: Would be nice to have a new line at the end of the file.

File: actions/node_delete.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+entry_point: 'node_delete.py'
+name: "node_delete"
+pack: "orion"
+runner_type: "python-script"
+
+parameters:
+  node:
+    type: "string"
+    description: The node to delete in Orion.
+    required: true
```
Comment (jjm) [inline actions/node_delete.yaml]: Would be nice to have a new line here too.

File: circle.yml [hunk 3]
```diff
@@ -19,7 +19,10 @@ test:
 
 deployment:
   index:
-    branch: master
+    branches:
```
Comment (jjm) [inline circle.yml]: @emedvedev @Kami May be you can comment on this? I think this change should be reversed before merging, as it should this only run for the master branch in the StackStorm-Exchange org.

**Comment by bishopbm1** [comment]
@LindsayHill Thank you for taking a look at this. I made the changes that you requested.

@jjm Thank you for also taking a look at this!

File: actions/lib/actions.py [hunk 4]
```diff
@@ -104,6 +104,36 @@ def get_node(self, node):
...
+        data = self.query(swql, **kargs)
+
+        if 'results' not in data:
+            msg = "No results from Orion: {}".format(data)
+            self.logger.info(msg)
+            raise Exception(msg)
+
+        if len(data['results']) == 1:
+            try:
+                orion_node.agent_id = data['results'][0]['AgentID']
```
Comment (jjm) [inline actions/lib/actions.py]: This agent logic should really be in `get_node` and `agent` (and the other values) are populated without needing to pass around the `orion_node` object.

File: actions/lib/node.py [hunk 5]
```diff
@@ -107,3 +110,31 @@ def uri(self):
...
+    @property
+    def agent_id(self):
+        return self._agent_id
+
+    @agent_id.setter
+    def agent_id(self, agent_id):
+        """
+        Set self._agent_id and update self._agent to True.
+        """
+        if agent_id is not None:
```
Comment (jjm) [inline actions/lib/node.py:123]: 👍

File: actions/lib/node.py [hunk 6]
```diff
@@ -107,3 +110,31 @@ def uri(self):
...
+            self._agent_id = agent_id
+            self._agent = True
+        else:
```
Comment (jjm) [inline actions/lib/node.py]: Is this `else` needed? They already default to these values.

File: actions/node_delete.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from lib.actions import OrionBaseAction
+
+
+class NodeDelete(OrionBaseAction):
```
Comment (jjm) [inline actions/node_delete.py:19]: Nice to see this action added.

File: actions/agent_delete.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        """
+
+        self.connect()
+
+        orion_node = self.get_node(node)
+
+        if not orion_node.npm:
+            raise ValueError("Node not found")
+
+        orion_node = self.get_agent(orion_node)
```
Comment (jjm) [inline actions/agent_delete.py]: See comments below about passing round the `orion_node` object.

File: actions/lib/actions.py [hunk 9]
```diff
@@ -104,6 +104,36 @@ def get_node(self, node):
 
         return orion_node
 
+    def get_agent(self, orion_node):
```
Comment (jjm) [inline actions/lib/actions.py]: I would prefer if this function took `npm_id` as a parameter, instead of `orion_node`.

**Review by jjm — CHANGES_REQUESTED** [review]
Overall very nice work, I would like to see slightly less code  around getting the `agent_id` duplicated in the actions instead of when getting an `orion_node` object.

With those small changes (and clarification on the `circle.yml changes), I'm very happy to merge this in.

**Comment by bishopbm1** [comment]
@jjm I made all the changes you suggested. I also changed the circle.yml back to only build the master branch.

**Review by jjm — APPROVED** [review]
LGTM

**Comment by jjm** [comment]
@bishopbm1 Thank you for the contrition!

**Pull request merged**